File: src/models/connection.ts
import type { DatabaseType } from "./database-type";
import { DATABASE_DEFAULTS } from "./database-type";

export type ConnectionColor =
  | "none" | "red" | "orange" | "yellow" | "green"
  | "blue" | "purple" | "pink" | "gray";

export const CONNECTION_COLORS: Record<ConnectionColor, string> = {
  none: "transparent",
  red: "#FF3B30",
  orange: "#FF9500",
  yellow: "#FFCC00",
  green: "#34C759",
  blue: "#007AFF",
  purple: "#AF52DE",
  pink: "#FF2D55",
  gray: "#8E8E93",
};

export type SSHAuthMethod = "password" | "private-key" | "ssh-agent";
export type SSLMode = "disabled" | "preferred" | "required" | "verify-ca" | "verify-identity";

export interface SSHConfiguration {
  enabled: boolean;
  host: string;
  port: number;
  username: string;
  authMethod: SSHAuthMethod;
  password: string;
  privateKeyPath: string;
  passphrase: string;
  agentSocketPath: string;
}

export interface SSLConfiguration {
  mode: SSLMode;
  caCertificatePath: string;
  clientCertificatePath: string;
  clientKeyPath: string;
}

export interface DatabaseConnection {
  id: string;
  name: string;
  type: DatabaseType;
  host: string;
  port: number;
  database: string;
  username: string;
  color: ConnectionColor;
  tagId: string | null;
  groupId: string | null;
  usePgpass: boolean;
  sshConfig: SSHConfiguration;
  sslConfig: SSLConfiguration;
}

export function createDefaultSSHConfig(): SSHConfiguration {
  return {
    enabled: false,
    host: "",
    port: 22,
    username: "",
    authMethod: "password",
    password: "",
    privateKeyPath: "",
    passphrase: "",
    agentSocketPath: "",
  };
}

export function createDefaultSSLConfig(): SSLConfiguration {
  return {
    mode: "disabled",
    caCertificatePath: "",
    clientCertificatePath: "",
    clientKeyPath: "",
  };
}

export function createDefaultConnection(type: DatabaseType = "postgresql"): DatabaseConnection {
  const defaults = DATABASE_DEFAULTS[type];
  return {
    id: crypto.randomUUID(),
    name: "",
    type,
    host: "localhost",
    port: defaults.defaultPort,
    database: "",
    username: "",
    color: "none",
    tagId: null,
    groupId: null,
    usePgpass: false,
    sshConfig: createDefaultSSHConfig(),
    sslConfig: createDefaultSSLConfig(),
  };
}

File: src-tauri/src/models/connection.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum DatabaseType {
    Postgresql,
    Mysql,
    Mariadb,
    Sqlite,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ConnectionColor {
    None,
    Red,
    Orange,
    Yellow,
    Green,
    Blue,
    Purple,
    Pink,
    Gray,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum SSHAuthMethod {
    Password,
    PrivateKey,
    SshAgent,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum SSLMode {
    Disabled,
    Preferred,
    Required,
    VerifyCa,
    VerifyIdentity,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SSHConfiguration {
    pub enabled: bool,
    pub host: String,
    pub port: u16,
    pub username: String,
    pub auth_method: SSHAuthMethod,
    #[serde(skip_serializing, default)]
    pub password: String,
    pub private_key_path: String,
    #[serde(skip_serializing, default)]
    pub passphrase: String,
    pub agent_socket_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SSLConfiguration {
    pub mode: SSLMode,
    pub ca_certificate_path: String,
    pub client_certificate_path: String,
    pub client_key_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatabaseConnection {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    pub db_type: DatabaseType,
    pub host: String,
    pub port: u16,
    pub database: String,
    pub username: String,
    pub color: ConnectionColor,
    pub tag_id: Option<String>,
    pub group_id: Option<String>,
    pub use_pgpass: bool,
    pub ssh_config: SSHConfiguration,
    pub ssl_config: SSLConfiguration,
}
